src/adapters/__tests__/commonTests.js

```diff
@@ -451,15 +452,162 @@ export default () => [
+        { name: 'test_boolean_optional', type: 'bool', isOptional: true },
+      ]
+      const projectColumnsV5 = [{ name: 'text2', type: 'string', isIndexed: true }]
+      const tagAssignmentSchema = {
+        name: 'tag_assignments',
+        columns: [{ name: 'text1', type: 'string' }],
+      }
+
+      const testSchemaV5 = appSchema({
+        version: 5,
+        tables: [
+          tableSchema({
+            name: 'tasks',
+            columns: [...taskColumnsV3, ...taskColumnsV5],
+          }),
+          tableSchema({
+            name: 'projects',
+            columns: [...projectColumnsV3, ...projectColumnsV5],
+          }),
+          tagAssignmentSchema,
+        ],
+      })
+      const migrationsV5 = schemaMigrations({
+        minimumVersion: 2,
+        currentVersion: 5,
+        migrations: [
+          {
+            from: 4,
+            to: 5,
+            steps: [addColumns({ table: 'tasks', columns: taskColumnsV5 })],
+          },
+          {
+            from: 3,
+            to: 4,
+            steps: [
+              createTable(tagAssignmentSchema),
+              addColumns({ table: 'projects', columns: projectColumnsV5 }),
+            ],
+          },
+          {
+            from: 2,
+            to: 3,
+            steps: [
+              createTable({
+                name: 'will_not_be_created',
+                columns: [{ name: 'num1', type: 'number' }],
+              }),
+            ],
+          },
+        ],
+      })
+      adapter = adapter.testClone({
+        schema: testSchemaV5,
+        migrationsExperimental: migrationsV5,
+      })
+
+      // check that the data is still there
+      expect(await adapter.count(new Query({ modelClass: MockTask }, []))).toBe(2)
+
+      // check if new columns were populated with appropriate default values
+      const checkTaskColumn = (columnName, expectedValue) =>
+        new Query({ modelClass: MockTask }, [Q.where(columnName, expectedValue)])
+
+      expect(await adapter.count(checkTaskColumn('test_string', ''))).toBe(2)
+      expect(await adapter.count(checkTaskColumn('test_string_optional', null))).toBe(2)
+      expect(await adapter.count(checkTaskColumn('test_number', 0))).toBe(2)
+      expect(await adapter.count(checkTaskColumn('test_number_optional', null))).toBe(2)
+      expect(await adapter.count(checkTaskColumn('test_boolean', false))).toBe(2)
+      expect(await adapter.count(checkTaskColumn('test_boolean_optional', null))).toBe(2)
+
+      // check I can use new table and columns
+      adapter.batch([
+        ['create', new MockTagAssignment({}, { id: 'tt2', text1: 'hello' })],
+        ['create', new MockProject({}, { id: 'p1', text1: 'hey', text2: 'foo' })],
+        [
+          'create',
+          new MockTask(
+            {},
+            { id: 't3', test_string: 'hey', test_number: 2, test_boolean_optional: true },
+          ),
+        ],
+      ])
+
+      // check that out-of-range migration was not executed
```

Comment:
Shouldn't out of range migration throw an error?

Reply:
Not sure what you're asking — the point is that database we're testing is already at version 3, so migrations from v2 to v3 must not be executed (and we're testing that the migration was not in fact applied)

Reply:
OK, my bad :)